charts/IndicatorDropdown.tsx

```diff
@@ -0,0 +1,131 @@
+import * as React from "react"
+import AsyncSelect from "react-select/async"
+import { ValueType, SingleValueProps } from "react-select"
+import { bind } from "decko"
+
+import {
+    ExplorerViewContext,
+    ExplorerViewContextType
+} from "./ExplorerViewContext"
+import { observable, action, runInAction } from "mobx"
+import { Indicator } from "./Indicator"
+import { observer } from "mobx-react"
+
+export interface IndicatorDropdownProps {
+    placeholder: string
+    selectedId?: number
+    onChangeId: (id: number) => void
+}
+
+@observer
+export class IndicatorDropdown extends React.Component<IndicatorDropdownProps> {
+    static contextType = ExplorerViewContext
+    context!: ExplorerViewContextType
+
+    static defaultProps = {
+        placeholder: "Type to search..."
+    }
+
+    @observable.ref currentIndicator?: Indicator | null
+
+    componentDidMount() {
+        this.loadCurrentIndicator()
+    }
+
+    componentDidUpdate() {
+        this.loadCurrentIndicator()
+    }
+
+    // Since this component only takes IDs as parameters, we need to load the
+    // full indicator in order to display the title.
+    @action async loadCurrentIndicator() {
+        if (
+            this.props.selectedId &&
+            this.props.selectedId !==
+                (this.currentIndicator && this.currentIndicator.id)
+        ) {
+            const fetchedId = this.props.selectedId
+            const indicator = await this.context.indicatorStore.get(
+                this.props.selectedId
+            )
+            // this.props.selectedId could've changed during the await, need to
+            // make a second check.
```

Comment:
I'm not totally sure what's causing this “could have changed during the wait” stuff, but this is the second place it's shown up so far. Seems awkward and error-prone. Why is this needed? Can we re-architect or re-design somehow to eliminate this?

Reply:
It's very fiddly, I agree it should be handled differently.

It's to handle this case:

```
switches to indicator 1 ---> waits impatiently ---> switches to indicator 2 ---> waits ---> indicator 2 shown ---> indicator 1 shown
                          \                      \                                      /                      /
                           \                      -- indicator 2 loading ------ loaded -                      /
                            \                                                                                /
                              -- indicator 1 loading ----------------------------------------------- loaded--
                                                  X at this point, indicator 1 should be cancelled
                                                  but promises are not cancellable
```

Indicator 1 gets delayed but its handler is still called last, so the current indicator is Indicator 1 despite the user choosing Indicator 2.

Reply:
I will work on this, I agree it's a problem

Reply:
OK. We can discuss, I probably have ideas

Reply:
So I think a more elegant solution, that would be more in line with the reactive style that React/mobx are meant for, would be to have `currentIndicator` not be an observable attribute, but a computed property that looks up the indicator in the store by id. That should be a *synchronous* method that returns either the indicator, if it's loaded, or an empty object or other placeholder value, if not. The currently selected ID, itself, should be an observable property, so that when it changes, it triggers `currentIndicator` to recompute.

Then there just needs to be some mechanism whereby the indicator store also retriggers the computation when it has loaded the data. That should be straightforward using mobx if we do this right.

Does that make sense? Basically, be more strict about the rendering being a pure computation derived from first principles.

Reply:
Curious to hear your thinking on rolling our own store vs. using an existing one, too, since I think that's relevant here… using a good one would make some of these basic architecture/design choices for us and maybe smooth the way if we find a good impedance match with what we're doing.

Reply:
Agree with exactly your proposal in the above comment re synchronous method!

Re switching to Redux: it comes with "declarative actions" and they are kind of the immutable solution to the problem mobx tries to solve with mutation. The two approaches might not work very well together I think, things might not render when they should, or re-render when they don't need to...

Reply:
Ah, interesting. Is there another model store that would be a better fit?

Reply:
Mobx has some recommendations here, we should both review this: https://mobx.js.org/best/store.html

Reply:
And there's a bunch of relevant things here, including a whole category of model libraries: https://github.com/mobxjs/awesome-mobx

Reply:
For mobx probably this: https://github.com/mobxjs/mobx-state-tree
Docs: https://mobx-state-tree.js.org/intro/philosophy

Haven't investigated much yet

Reply:
The store issue ended up being quite fiddly. In order to be able to return “placeholder” values while the real indicator is loading, we could: 

1. Create our own getter function (`get()`) that creates and returns the placeholder value and at the same time schedules a fetch to populate the store with the real value.
2. Create a `fetch()` function that populates the store with a placeholder value and schedules a fetch request, but then retrieve the placeholder value with property access, e.g. `indicators[indicatorId]`.

The first option is problematic because [mobx tracks property access, not values](https://mobx.js.org/best/react.html#mobx-tracks-property-access-not-values), so observers that call `get()` but do nothing else will not automatically subscribe to value changes. I chose this option and that's why there's a `noop()` call in ExploreView.

The second option seems bad because we always have to remember to call load before accessing, but then we don't have the problems with property access.

Reply:
One way I have seen this handled is like this:

* The first time you call store.get(id), an empty model object is created with that id, put in the store, and returned to you
* At the same time, the store makes a fetch call for the data
* When the data comes back, the model object is updated. So if the view is reacting to its properties, it re-renders

That solves the problem with option 1 and it's what I'd be inclined to do. There could be other ways to solve this, but I like this pattern in general.